lib/xwax/pitch_kalman.h

```diff
@@ -0,0 +1,325 @@
+
+    /*
+     * Predict covariance: P_pred = F * P * F^T + Q
+     * For 2×2 P and F, expand each term explicitly:
+     *
+     *   P_pred[x][x] = F00 * (F00 * P[x][x] + F01 * P[x][v]) + F01 * (F00 * P[x][v] + F01 * P[v][v]) + Q11
+     *   P_pred[x][v] = F10 * (F00 * P[x][x] + F01 * P[x][v]) + F11 * (F00 * P[x][v] + F01 * P[v][v]) + Q12
+     *   P_pred[v][v] = F10 * (F10 * P[x][x] + F11 * P[x][v]) + F11 * (F10 * P[x][v] + F11 * P[v][v]) + Q22
+     *
+     * This preserves symmetry while avoiding full matrix multiplication.
+     */
+
+    const double P[2][2] = { {p->P[x][x], p->P[x][v]},
+                             {p->P[v][x], p->P[v][v]} };
+
+    double P_pred[2][2];
+
+    P_pred[x][x] = F[x][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[x][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][x];
+
+    P_pred[x][v] = F[v][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[v][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][v];
+
+    P_pred[v][v] = F[v][x] * (F[v][x] * P[x][x] + F[v][v] * P[x][v]) +
+                   F[v][v] * (F[v][x] * P[x][v] + F[v][v] * P[v][v]) + Q[v][v];
+
+    /*
+     * =================== Update step =======================
+     *
+     *   Measurement model: z = Hx + noise
+     *   Here H = [1 0], so z = x + noise
+     *
+     *   Innovation (residual): y = z - H * x_pred
+     *   Since Hx = 1, Hv = 0, this reduces to: y = dx - x_pred
+     */
+
+    const double Hx = 1.0, Hv = 0.0;
+    const double y = dx - (Hx * x_pred + Hv * v_pred);
+
+    /*
+     * Toggle mode switches for stable playback and scratching. When the
+     * innovation quantity hits a certain treshold the filter sensitivity
+     * is tuned.
+     *
+     * The innovation is similar to the residual in the alpha-beta filter,
+     * calculated as
+     *
+     *          residual = measurement - prediction
+     *
+     * It represents how much the actual measurement deviates from the
+     * predicted value.
+     *
+     * NOTE: Probably better to use a continuous curve here instead of switches.
+     */
+
+    const double y_abs = fabs(y);
+
+    if (y_abs > p->scratch_threshold)
+        kalman_tune_sensitivity(p, &p->reactive);
+    else if (y_abs > p->medium_threshold && y_abs < p->scratch_threshold)
+        kalman_tune_sensitivity(p, &p->medium);
+    else
+        kalman_tune_sensitivity(p, &p->stable);
+
+    /*
+     * Innovation covariance: S = H * P_pred * H^T + R
+     * With H = [1 0], this reduces to: S = P_pred[x][x] + R
+     */
+
+    const double S = P_pred[x][x] + p->coeffs->R;
+
+    /*
+     * Kalman gain: K = P_pred * H^T * S^-1
+     *
+     * With H = [1 0]:
+     *   Kx = P_pred[x][x] / S
+     *   Kv = P_pred[x][v] / S
+     */
+
+    const double invS = 1.0 / S;
+    const double Kx = P_pred[x][x] * invS;
+    const double Kv = P_pred[x][v] * invS;
+
+    /*
+     * Update state:
+     *
+     * x_upd = x_pred + Kx * y
+     * v_upd = v_pred + Kv * y
+     */
```

Comment:
not super useful IMO, rephrase in terms of the more abstract matrix operations this is modelling?

```suggestion
    /*
     * Update state:
     *
     * xv_upd = xv_pred + K * y
     */
```

Reply:
It's a draft